homeassistant/components/media_player/yamaha.py

```diff
@@ -49,8 +57,40 @@ def update(self):
             self._pwstate = STATE_OFF
         self._muted = self._receiver.mute
         self._volume = (self._receiver.volume/100) + 1
-        self._current_source = self._receiver.input
-        self._source_list = list(self._receiver.inputs().keys())
+
+        if self.source_list is None:
+            self.build_source_list()
+
+        current_source = self._receiver.input
+        if self._mapping is not None:
+            if current_source in list(self._mapping.keys()):
+                self._current_source = self._mapping[current_source]
+            else:
+                self._current_source = current_source
+        else:
+            self._current_source = current_source
+
+    def build_source_list(self):
+        """Build the source list."""
+        self._source_list = []
+
+        if self._mapping is not None:
+            self._reverse_mapping = {}
+            mapping_list = list(self._mapping.keys())
+            for source, alias in self._mapping.items():
+                self._reverse_mapping[alias] = source
+        else:
+            mapping_list = []
+
+        source_list = list(self._receiver.inputs().keys())
+        for source in source_list:
+            if self._excludes is not None and source in self._excludes:
+                continue
+            if source in mapping_list:
+                self._source_list.append(self._mapping[source])
+            else:
+                self._source_list.append(source)
+        self._source_list.sort()
```

Comment:
If you make sure `self._excludes` is always an empty list if none given, you can do this:
```python
self._source_list = sorted(self._mapping.get(source, source) for source in self._receiver.inputs()
                           if source not in self._excludes)
```